furnace-service-provider/src/main/java/org/jboss/windup/web/furnaceserviceprovider/FurnaceExtension.java

```diff
@@ -0,0 +1,392 @@
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.Default;
+import javax.enterprise.inject.spi.AfterBeanDiscovery;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.enterprise.inject.spi.InjectionTarget;
+import javax.enterprise.util.AnnotationLiteral;
+
+import org.jboss.forge.furnace.Furnace;
+import org.jboss.forge.furnace.addons.Addon;
+import org.jboss.forge.furnace.addons.AddonStatus;
+
+/**
+ * @author <a href="mailto:jesse.sightler@gmail.com">Jesse Sightler</a>
+ */
+public class FurnaceExtension implements Extension
+{
+    private static Logger LOG = Logger.getLogger(FurnaceExtension.class.getName());
+
+    public static final String WINDUP_WEB_SUPPORT = "windup-web-support";
+    public static final String PACKAGE_PREFIX = "org.jboss.windup.web";
+
+    private FurnaceProducer furnaceProducer;
+
+    private FurnaceProducer getFurnaceProducer()
+    {
+        if (furnaceProducer == null)
+        {
+            synchronized (FurnaceExtension.class)
+            {
+                FurnaceProducer furnaceProducer = new FurnaceProducer();
+                furnaceProducer.setup(WebProperties.getInstance().getAddonRepository());
+                this.furnaceProducer = furnaceProducer;
+            }
+        }
+        return furnaceProducer;
+    }
+
+    public void afterBeanDiscovery(@Observes AfterBeanDiscovery abd, BeanManager beanManager)
+    {
+        FurnaceProducer furnaceProducer = getFurnaceProducer();
+        addFurnace(abd, beanManager, furnaceProducer);
+
+        HashSet<String> duplicateCheck = new HashSet<>();
+        for (Addon addon : furnaceProducer.getFurnace().getAddonRegistry().getAddons())
+        {
+            if (addon.getId() != null && addon.getStatus() == AddonStatus.STARTED && addon.getId().toString().contains(WINDUP_WEB_SUPPORT))
+            {
+                for (Class<?> exportedType : addon.getServiceRegistry().getExportedTypes())
+                {
+                    addService(duplicateCheck, abd, beanManager, exportedType);
+                }
+            }
+        }
+    }
+
+    private <T> void addService(final Set<String> duplicateCheck, final AfterBeanDiscovery abd, final BeanManager beanManager, final Class<T> incomingType)
+    {
+        final FurnaceProducer furnaceProducer = getFurnaceProducer();
+        if (incomingType.getPackage() == null || !incomingType.getPackage().getName().startsWith(PACKAGE_PREFIX))
+            return;
+
+        if (duplicateCheck.contains(incomingType.getCanonicalName()))
+            return;
+
+        duplicateCheck.add(incomingType.getCanonicalName());
+
+        final Class<T> type;
+        boolean registerWithWeld = true;
+
+        // Try to make sure that we are using the one from the local classloader instead of a Furnace addon classloader
```

Comment:
Because otherwise the class.equals() check may fail (as the classes may come from two different classloaders).